### PR TITLE
Implement Relevance Functions

### PR DESCRIPTION
**Pull request by beccadaniel**
## Why make this change?
We need to add relevance functions to our vector store implementation. This includes enabling users to be able to provide the relevance function they want and subsequently run similarity searches with relevance scores.

## What is this change?
This PR includes implementation for the following functions:
- `_select_relevance_score_fn` - This function determines which relevance function will be used when `similarity_search_with_relevance_scores` is invoked. The user can either provide a relevance function or we choose a relevance function based on the selected distance strategy. The implementation for `similarity_search_with_relevance_scores` and the different relevance functions that correspond to `COSINE`, `DOT` & `EUCLIDEAN` are available in LangChain code.- 
- `max_marginal_relevance_search` - Uses `maximal_marginal_relevance` implementation from LangChain code.
- `max_marginal_relevance_search_by_vector`
- `_search_store` helper function has been updated to return results that can be processed for maximal marginal relevance function. This result differs from the original implementation by making an extra query to get the embedding.

## Tests
The test suite has been updated to use `DeterministicFakeEmbeddings` instead of `FakeEmbeddings`. The difference between these two classes is that `DeterministicFakeEmbeddings` uses a seed for generating embeddings and thus makes it better for debugging purposes at a later date.

image

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/sqlserver.py [hunk 1]
```diff
@@ -331,6 +337,111 @@ def from_texts(
...
+            embedding_list,
+            lambda_mult=lambda_mult,
+            k=k,
+        )
+
+        results_as_docs = self._docs_from_result(
+            self._docs_and_scores_from_result(results)
+        )
+
+        # Return list of Documents from results_as_docs_and_scores whose position
```
Comment (Aniruddh25) [inline libs/community/langchain_community/vectorstores/sqlserver.py]:
```suggestion
        # Return list of Documents from results_as_docs whose position
```

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py [hunk 2]
```diff
@@ -594,6 +594,37 @@ def test_that_entra_id_authentication_connection_is_successful(
...
+    """Test that the size of documents returned when
+    `similarity_search_with_relevance_scores` is called
+    is the expected number of documents requested."""
+    number_of_docs_to_return = 3
+
+    store.add_texts(texts, metadatas)
+    result = store.similarity_search_with_relevance_scores(
+        "Good review", k=number_of_docs_to_return
+    )
+    assert len(result) == number_of_docs_to_return
```
Comment (Aniruddh25) [inline libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:717]: this is only testing we get the expected number of docs. Can we also check the result is constrained between 0 and 1. 

Also, is there any way to test if the results we are getting are as expected as per maximal_marginal_relevance? Meaning that those would be different if we were to NOT use such a relevance_score?

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py [hunk 3]
```diff
@@ -594,6 +594,37 @@ def test_that_entra_id_authentication_connection_is_successful(
...
+
+
```
Comment (Aniruddh25) [inline libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:719]: another test to be added is to see if we pick the right maximal_marginal_relevance strategy based on the distance function. or spit out an error to provide the relevance function.

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py (hunk 2 above)
Comment (beccadaniel) [reply libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:717]: Our results here aren't always constrained between 0 and 1. That's what @yorek will be looking into.
If we do this, we'll be introducing flakiness to this test.

Yes, we can determine expected value with the formula -> (1 -`similarity_search_score`). And check that the result from the call to `similarity_search_with_relevance_scores` is expected.

Comment (yorek) [reply libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:717]: I checked the test, and the issue is the vector are randomly generated by the `FakeEmbeddings` function. I tried using vectors with only three dimensions and the generated vectors where:

v1 = [-2.3745856e+000,1.2014049e+000,6.8701744e-001]
v2 = [5.9738564e-001,3.5180563e-001,4.1327056e-001]

if you calculate the cosine distance the result is 1.32094580014782 which means that the resulting similarity (or score) is `1-1.32094580014782` = -0.32094580014782

Here's the manual proof:

```sql
with 
v1 as (select cast(value as float) as v, * from openjson('[-2.3745856e+000,1.2014049e+000,6.8701744e-001]')),
v2 as (select cast(value as float) as v,* from openjson('[5.9738564e-001,3.5180563e-001,4.1327056e-001]'))
select
 1-SUM(v1.v * v2.v) / 
    (
        SQRT(SUM(v1.v * v1.v)) 
        * 
        SQRT(SUM(v2.v * v2.v))
    ) as cosine_distance
from
v1 inner join v2 on v1.[key] = v2.[key]
```

I think this issue can be ignored for now. I would suggest opening an issue on the langchain repo to ask for confirmation and/or propose to use only positive values for generating vectors, after making that change, test was running green all the time, while before it was very inconsistent (as expected given the vectors are completely random)/

This is the change I did in `fake.py` file and added `abs` going from

```python
    def _get_embedding(self) -> List[float]:
        return list(np.random.normal(size=self.size))
```

to

```python
    def _get_embedding(self) -> List[float]:
        return list(abs(np.random.normal(size=self.size)))
```

to make sure generated vectors are always positive and the cosine distance is between 0 and 1

Comment (yorek) [reply libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:717]: Actually, I submitted a PR already :) https://github.com/langchain-ai/langchain/pull/27406

**Review by yorek — APPROVED** [review]

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py [hunk 4]
```diff
@@ -713,6 +717,37 @@ def test_similarity_search_with_relevance_score(
...
+) -> None:
+    """Test that `_select_relevance_score_fn` uses the appropriate
+    relevance function based on distance strategy provided."""
+    store.add_documents(docs)
+    result = store._select_relevance_score_fn()
+
+    # This vectorstore is initialized to use `COSINE` distance strategy,
+    # as such, the relevance function returned should be a
+    # cosine relevance function.
+    assert result == store._cosine_relevance_score_fn
```
Comment (Aniruddh25) [inline libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:748]: can we test if we give error if the distance strategy is not one of the supported ones?

**Review by Aniruddh25 — APPROVED** [review]
LGTM

File: libs/community/tests/unit_tests/vectorstores/test_sqlserver.py (hunk 4 above)
Comment (beccadaniel) [reply libs/community/tests/unit_tests/vectorstores/test_sqlserver.py:748]: Added a test for that.

**Review by Aniruddh25 — APPROVED** [review]

**Pull request merged**